### PR TITLE
added fromJSON()

### PR DESCRIPTION
**Pull request by bbonczek**
Concerns issue #31 Added method allowing for deserializing to Midi object.

List of changes:
a) added `deserialize(json)` method to `MidiConvert`
b) added `fromJSON(json)` method to `Midi, Track, Note`
c) updated tests & d.ts definitions

### CODE REVIEW

**Comment by bbonczek** [comment]
Hi, i guess that's it. Should close it now, or what? I't my first pull request.

File: src/Midi.js [hunk 1]
```diff
@@ -161,6 +161,21 @@ class Midi {
 	}
 
 	/**
+	 * Convert JSON to Midi object
+	 * @param {object} json
+	 * @returns {Midi}
+	 */
+	fromJSON(json){
```
Comment (appsforartists) [inline src/Midi.js]: I would make this a static method.  (`encode` probably should be too, but that's a separate conversation).

File: src/MidiConvert.js [hunk 2]
```diff
@@ -66,6 +66,14 @@ const MidiConvert = {
 	 */
 	create : function(){
 		return new Midi()
+	},
+	/**
+	 * Create midi object from json
+	 * @param {object} json
+	 * @returns {Midi} Deserialized midi object
+	 */
+	deserialize : function(json){
```
Comment (appsforartists) [inline src/MidiConvert.js]: Let's also call this `fromJSON`

File: test/test.js [hunk 3]
```diff
@@ -336,6 +371,19 @@ describe("Note", function(){
 		expect(json.name).to.equal("test")
 	})
 
+	it("can decode track from JSON", function(){
```
Comment (appsforartists) [inline test/test.js]: Does this test anything the other one you added doesn't?

File: src/Track.js [hunk 4]
```diff
@@ -317,6 +317,31 @@ class Track {
 		return ret
 	}
 
+	/**
+	 * Convert JSON to Track object
+	 * @param {object} json
+	 * @returns {Track}
+	 */
+	fromJSON(json){
```
Comment (appsforartists) [inline src/Track.js]: This should also be `static` (which means it should go at the top of the class).

**Review by appsforartists — CHANGES_REQUESTED** [review]

**Comment by appsforartists** [comment]
@BartoszBaczek Please make the requested changes here.

If you change your summary from "Concerns #31" to "Closes #31", your issue will automatically close when this lands.

**Comment by appsforartists** [comment]
Also, the name of this PR will be the name of your commit, so I'd change it to **"added fromJSON()"**

File: test/test.js (hunk 3 above)
Comment (bbonczek) [reply test/test.js]: You're right, those are already tested in 'is jSON deserializable'

**Comment by bbonczek** [comment]
Done.

**Comment by appsforartists** [comment]
Great!

Your static methods should be at the very top of their classes, before their constructors:

```typescript
class Midi {
  static fromJSON(json) {}
  
  constructor() {}
}
```

**Comment by bbonczek** [comment]
Done :) May I ask, when are you planning to bump npm package?

File: src/MidiConvert.d.ts [hunk 5]
```diff
@@ -48,6 +48,7 @@ export interface MIDI {
 export function parse(raw: ArrayBuffer): MIDI;
 export function load(url: string, data?: any, method?: 'GET'|'POST'): Promise<MIDI>;
 export function create(): MIDI;
+export function fromJSON(json: object): MIDI;
```
Comment (appsforartists) [inline src/MidiConvert.d.ts:51]: I wonder if `Partial<MIDI>` is a more accurate type for `json`.  `json` should have the same shape as `MIDI`, just without the methods, right?

I'm not sure if `Partial` is deep though, e.g. you'd want tracks to also be `Partial<Track>`, and I don't think `Partial<MIDI>` gets you that by default.

**Review by appsforartists — APPROVED** [review]

**Comment by appsforartists** [comment]
I'm not an owner of this repo, so I can't comment on that.  You'd have to ask @tambien.

**Comment by appsforartists** [comment]
One more thing: the first message in this thread will become the commit message, so you should update it:

> Added static `fromJSON` method to the index and to each of `Midi`, `Note`, and `Track`.
> 
> Closes #31

**Comment by tambien** [comment]
thank you both!

**Pull request merged**